dashboard/react-components: robot-page

vallq opened the pull request:
## What's new
Robot Panel component based off from Task Panel
![robot panel ui](https://user-images.githubusercontent.com/28668944/118419129-bd0a3800-b6ed-11eb-8946-eef61093b1dc.png)
https://user-images.githubusercontent.com/28668944/118747937-866b2380-b88d-11eb-9fae-ccff90669be0.mp4
addition of robot page to the dashboard

## Self-checks

- [x] I'm familiar with and follow this [Typescript guideline](https://basarat.gitbook.io/typescript/styleguide)
- [x] I added unit-tests for new components
- [ ] I tried testing edge cases
- [ ] I tested the behavior of the components that interact with the backend, with an e2e test

## Discussion
1. I'm still not sure how to use the `api_server`'s implementation of task progress in `react-components` because what react-components use is technically just dummy data. should this be taken as a prop and only implemented in the dashboard?

vallq commented:
@koonpeng @matiasbavera hello, this is the implementation of the robots page. my current question is as described above, but I think the rest API should be able to resolve that? not entirely sure.

also I originally wanted to use antd's progress bar component to indicate task progress status because mui does not have such a component. but it would be "overkill" to include this, right? was thinking that a number would suffice.
![tasks_v1](https://user-images.githubusercontent.com/28668944/118748929-28d7d680-b88f-11eb-810e-69d43d3b85f5.png)

matiasbavera commented:
The truth is that this is neat but it seems asymmetrical. For testing purposes, what do you think about putting `Battery` and` Assigned Tasks` centered to see how it looks

![image](https://user-images.githubusercontent.com/11761240/118876764-a0235f80-b8bb-11eb-81eb-ca9e6eaf4c54.png)

koonpeng commented:
> I'm still not sure how to use the api_server's implementation of task progress in react-components because what react-components use is technically just dummy data. should this be taken as a prop and only implemented in the dashboard?

1 way is to `import type {} from 'api-client`, another is to just take in a string, the dashboard will handle extracting the correct data from the api response.

vallq commented:
@matiasbavera I'm not sure about the GH link you shared because it uses the original material UI library and it also uses `styled-components` which is something that we do want to remove, right?
I did find another: https://github.com/martyan/react-customizable-progressbar, which is also more updated. just that there's some webpack version incompatibility if I run the install line. will try this out and see!

@koonpeng ooo okay, will try that, thank you

koonpeng commented on packages/dashboard/src/components/robots/robot-page.tsx:
```diff
@@ -0,0 +1,76 @@
+    if (!tasksApi) {
+      return;
+    }
+    const tasks = await tasksApi.getTasksTasksGetTasksGet();
```
In order to save bandwidth, we need to add a robot query api to the server. The returned result would include the current task of each robot. It will be similar to the new tasks query api added in #341.

I can help add the new api, is there any other data that will be needed?

vallq replied:
oh okay!
other than the current tasks, is it possible to retrieve all assigned and active tasks? (completed/failed tasks are not needed). because there is an 'assigned tasks' section in the robot-info.tsx

vallq commented:
Original left alignment
![left align](https://user-images.githubusercontent.com/28668944/119070673-0f639580-ba1b-11eb-9765-be6721731100.png)
Centered battery and assigned tasks
![centered](https://user-images.githubusercontent.com/28668944/119070730-24402900-ba1b-11eb-97f0-3b96add723bf.png)
The battery bar is already centered but it looks otherwise because there is white space after the '%'
![alignment](https://user-images.githubusercontent.com/28668944/119071135-d7a91d80-ba1b-11eb-9c68-3aa1696f55ed.png)
Another option I can think of, but might get a little messy?
![image](https://user-images.githubusercontent.com/28668944/119072108-8d28a080-ba1d-11eb-8367-ec7126f52411.png)

matiasbavera commented:
> Another option I can think of, but might get a little messy?
> ![image](https://user-images.githubusercontent.com/28668944/119072108-8d28a080-ba1d-11eb-8367-ec7126f52411.png)

Would it lose context if we do something like this?
``` 
               Battery

           Task progress

Assigned Task     Est. End time

Location              Destination
```

vallq commented:
I've addressed the comments made in the previous branch pr (`ui/robot-panel`). Thanks @matiasbavera and @koonpeng 
Some of the comments with 'VerboseRobot' updates, I think those can be made in another PR after this and the robot endpoints are merged in?

koonpeng commented on packages/react-components/lib/robots/utils.ts:
```diff
@@ -29,3 +30,58 @@ export function robotModeToString(robotMode: RmfModels.RobotMode): string {
+export interface VerboseRobot {
+  name: string;
+  model: string;
+  task_id: string;
```
should be `camelCase`.

koonpeng commented on packages/react-components/lib/robots/utils.ts:
```diff
@@ -29,3 +30,58 @@ export function robotModeToString(robotMode: RmfModels.RobotMode): string {
+  task_id: string;
+  seq: number;
+  mode: RmfModels.RobotMode;
+  battery_percent: number;
```
should be `camelCase`.

koonpeng commented on packages/react-components/lib/robots/utils.ts:
```diff
@@ -29,3 +30,58 @@ export function robotModeToString(robotMode: RmfModels.RobotMode): string {
+  battery_percent: number;
+  location: RmfModels.Location;
+  path: RmfModels.Location[];
+  assigned_tasks: TaskProgress[];
```
should be `camelCase`.

koonpeng commented on packages/react-components/lib/robots/utils.ts:
```diff
@@ -29,3 +30,58 @@ export function robotModeToString(robotMode: RmfModels.RobotMode): string {
 }
+
+//TODO: endpoint should return the equivalent of this
+export interface VerboseRobot {
```
If you mean to extend the base `RobotState` you can do
```tsx
export interface VerboseRobot extends RmfModels.RobotState {
```

koonpeng commented on packages/react-components/stories/circular-progress-bar.stories.tsx:
```diff
@@ -0,0 +1,37 @@
+import { Typography } from '@material-ui/core';
+
+export default {
+  title: 'Circular Progress Bar',
```
Move this to a `Robots` section if `ProgressBar` and `CircularProgressBar` is moved to be next to `RobotInfo`.

koonpeng commented on packages/react-components/stories/progressbar.stories.tsx:
```diff
@@ -0,0 +1,26 @@
+import { LinearProgressProps } from '@material-ui/core';
+
+export default {
+  title: 'Progress Bar',
```
Move this to a `Robots` section if `ProgressBar` and `CircularProgressBar` is moved to be next to `RobotInfo`.

koonpeng commented on packages/react-components/stories/progressbar.stories.tsx:
```diff
@@ -0,0 +1,26 @@
+  value: 100,
+};
+
+export const SecondaryProgressBar: Story<LinearProgressProps & { value: number }> = (args) => {
```
This is probably not needed as the values and colors can easily be changed with storybook settings.

koonpeng commented on packages/react-components/stories/circular-progress-bar.stories.tsx:
```diff
@@ -0,0 +1,37 @@
+  strokeColor: 'green',
+};
+
+export const StalledProgressBar: Story<CircularProgressBarProps> = (args) => {
```
This is probably not needed as the values and colors can easily be changed with storybook settings.

koonpeng commented on packages/react-components/tests/robots/robot-panel.spec.tsx:
```diff
@@ -0,0 +1,32 @@
+    const tasks = [makeDefinedTask('Loop', 'test_robot1', 'task_1', 3, 3)];
+    const robots = [makeRandomRobot('test_robot1', 'test_fleet', 2)];
+    const root = render(<RobotPanel robots={robots} fetchTasks={makeFetchTasks(tasks)} />);
+    root.debug();
```
I think this is leftover from testing?

koonpeng commented on packages/react-components/tests/test-data/tasks.ts:
```diff
@@ -24,3 +25,98 @@ export function makeTask(
+
+export function makeDefinedTask(
+  type: string,
+  robot_name: string,
```
camelCase

matiasbavera commented:
Hi, @vallq I'll approve it for you to merge it "today", so we can create a branch from it. But can you please address the minor changes pointed out from @koonpeng, test all the new features again, and create a follow-up GH issue of the things that we need to do?

vallq commented on packages/react-components/lib/robots/robot-info.tsx:
```diff
@@ -0,0 +1,222 @@
+          return task.task_summary.task_id;
+        }
+      })
+      .join('');
```
oh but this means that the last ID will also have an arrow
`1002 → 1004 → `
compared to `1002 → 1004`
